### PR TITLE
Noise: backbone logic for the Noise communication channel handshake

### PR DESCRIPTION
**Pull request by JSwambo**
Complete reimplementation of the handshake logic as an alternative to #5.

### CODE REVIEW

File: src/noise.rs [hunk 1]
```diff
@@ -336,8 +337,11 @@ pub fn encrypt_message(
 
     // Pad message
     // FIXME: padding is huge
-    let mut message_body = [0u8; NOISE_PADDED_MESSAGE_SIZE];
-    message_body.copy_from_slice(message);
+    let mut message_body = Vec::new();
+    message_body.extend_from_slice(message);
+    while message_body.len() < NOISE_PADDED_MESSAGE_SIZE {
+        message_body.extend_from_slice(&[0u8; 1]);
+    }
```
Comment (darosior) [inline src/noise.rs:344]: What was wrong with my version ? Here you heap-allocate and re-allocate `NOISE_PADDED_MESSAGE_SIZE - message_body.len()` times..

File: src/noise.rs [hunk 2]
```diff
@@ -383,3 +387,132 @@ pub fn decrypt_message(
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snow::{params::NoiseParams, Keypair};
+    use std::convert::TryInto;
+
+    /// Revault must specify the SodiumResolver to use sodiumoxide as the cryptography provider
+    /// when generating a static key pair for secure communication.
+    pub fn generate_keypair(noise_params: NoiseParams) -> Keypair {
```
Comment (darosior) [inline src/noise.rs:399]: nit: this could be a reference, would avoid an unnecessary `clone()` below.

File: src/noise.rs [hunk 3]
```diff
@@ -383,3 +387,132 @@ pub fn decrypt_message(
...
+
+        // client
+        let cli_act_2 = KXHandshakeActTwo::initiator(cli_act_1, &msg_2).unwrap();
+        let mut client_channel = KXChannel::from_handshake(cli_act_2).unwrap();
+
+        // client encrypts message for server
+        let msg = "Hello".as_bytes();
+        let encrypted_msg = encrypt_message(&mut client_channel, &msg).unwrap();
+        let decrypted_msg = decrypt_message(&mut server_channel, &encrypted_msg).unwrap();
+        assert_eq!(msg.to_vec(), decrypted_msg);
```
Comment (darosior) [inline src/noise.rs:435]: we def need a fuzz target for this. Will take care of that.

File: src/noise.rs [hunk 4]
```diff
@@ -383,3 +387,132 @@ pub fn decrypt_message(
...
+
+        // Test if padding hides message size
+        let msg_a = "".as_bytes();
+        let ciphertext_a = encrypt_message(&mut server_channel, &msg_a).unwrap();
+
+        let msg_b = [0u8; NOISE_MESSAGE_MAX_SIZE];
+        let ciphertext_b = encrypt_message(&mut server_channel, &msg_b).unwrap();
+
+        assert_eq!(ciphertext_a.0.len(), ciphertext_b.0.len());
+    }
```
Comment (darosior) [inline src/noise.rs:517]: Can you add a test which verifies we err if:
- we try to communicate without having terminated the handshake
- we mix up responder and server, both during and after the handshake
- w/e edge case you could think about :)

**Review by darosior — COMMENTED** [review]
Looks mostly good, just a few comments

File: src/noise.rs (hunk 3 above)
Comment (darosior) [reply src/noise.rs:435]: https://github.com/re-vault/revault_net/issues/9#issue-754287499

File: src/noise.rs (hunk 2 above)
Comment (JSwambo) [reply src/noise.rs:399]: `Builder::with_resolver()` moves noise_params so that doesn't work here

File: src/noise.rs (hunk 1 above)
Comment (JSwambo) [reply src/noise.rs:344]: `    let mut message_body = [0u8; NOISE_PADDED_MESSAGE_SIZE];
    message_body.copy_from_slice(message);`

You're solution panics so I just used one that worked. Just looked into it though and found the fix, which is that the slices need to be the same length: 

`    let mut message_body = [0u8; NOISE_PADDED_MESSAGE_SIZE];
    message_body[..message.len()].copy_from_slice(message);`

File: src/noise.rs (hunk 4 above)
Comment (JSwambo) [reply src/noise.rs:517]: I think its better if we change the fields to be private in `KXHandshakeActOne`, `KXHandshakeActTwo`, `KKHandshakeActOne`, `KKHandshakeActTwo`, `KXMessageActOne`, `KXMessageActTwo`, `KXMessageActOne` and `KXMessageActTwo`. Then the constructor methods are required and it's impossible to construct either `KXChannel` or `KKChannel` without completing a handshake. 

Also `NoiseEncryptedMessage' can have a private field.

File: src/noise.rs [hunk 5]
```diff
@@ -64,9 +74,11 @@ impl KXHandshakeActOne {
                     e
                 ))
             })?;
-        msg.truncate(len);
 
-        Ok((KXHandshakeActOne { state }, KXMessageActOne(msg)))
+        Ok((
+            KXHandshakeActOne { state },
+            KXMessageActOne(msg[..len].try_into().unwrap()),
```
Comment (darosior) [inline src/noise.rs:80]: Wow, any reason for the introduction of a panic ?...

**Review by darosior — COMMENTED** [review]
Just quickly skimmed the begining, will review tomorrow

File: src/noise.rs (hunk 5 above)
Comment (JSwambo) [reply src/noise.rs:80]: Well, there is some weird behaviour here from snow (see FIXME). If I can figure that out then there will be no need for using a Vec for the message, and no need to convert the vec into an array.

As it is now, I’m not sure if the try_into can error... at least for up-to-date rust versions.

File: src/noise.rs (hunk 4 above)
Comment (darosior) [reply src/noise.rs:524]: That's already the case, but fields aren't private (you need it in the next Act). The second and third points still hold though :)

File: src/noise.rs (hunk 1 above)
Comment (darosior) [reply src/noise.rs]: AH, good catch, thanks

File: src/test_utils.rs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    Ok(noise_params)
+}
+
+/// Revault must specify the SodiumResolver to use sodiumoxide as the cryptography provider
+/// when generating a static key pair for secure communication.
+pub fn generate_keypair(noise_params: NoiseParams) -> Keypair {
+    Builder::with_resolver(noise_params, Box::new(SodiumResolver::default()))
+        .generate_keypair()
+        .unwrap()
+}
```
Comment (darosior) [inline src/test_utils.rs:28]: This should be part of the `test` module.

File: src/noise.rs [hunk 7]
```diff
@@ -54,7 +62,9 @@ impl KXHandshakeActOne {
             .map_err(|e| Error::Noise(format!("Failed to build state for initiator: {:?}", e)))?;
 
         // Write the first message
-        let mut msg = vec![0u8; NOISE_MESSAGE_MAX_SIZE];
+        // FIXME: Why does the size here need to be larger than what is written?
+        // The write fails if msg.len() == KX_MSG_1_size but then len == KX_MSG_1_size
+        let mut msg = vec![0u8; KX_MSG_1_SIZE + MAC_SIZE];
```
Comment (darosior) [inline src/noise.rs:67]: That's the tag:
https://github.com/mcginty/snow/blob/ec32cfa89f026d1eccca27b2e7d0da8ef441bbcd/src/handshakestate.rs#L309-L320

However, a question remains: how comes that the tag isn't set (`[0; 16]` in our case) ? Even if a payload is passed, it's just copied as cleartext. So it seems that when we reach this:
https://github.com/mcginty/snow/blob/ec32cfa89f026d1eccca27b2e7d0da8ef441bbcd/src/symmetricstate.rs#L92-L107

We don't have an inner key.. But how do we set it ? Do we need to ? It seems to be missing the poly1305 tag here (at least, that's the exact size).

**Review by darosior — COMMENTED** [review]
Needs rebase. I'm still in the process of testing it but here is some more comments.

File: src/noise.rs (hunk 7 above)
Comment (darosior) [reply src/noise.rs:67]: Ok, so it seems to be normal. We don't use a PSK, thus we don't mix the keys and thus don't have a key for the symetric state and thus can't produce the tag:
https://github.com/mcginty/snow/blob/ec32cfa89f026d1eccca27b2e7d0da8ef441bbcd/src/handshakestate.rs#L243-L245

File: src/noise.rs [hunk 8]
```diff
@@ -491,28 +499,80 @@ mod tests {
...
+        let bad_msg = KKMessageActTwo([1u8; KK_MSG_2_SIZE]);
+        let cli_act_2 = KKHandshakeActTwo::initiator(cli_act_1, &bad_msg);
+        assert!(cli_act_2.is_err());
+
+        // KX handshake fails on client side if messages are badly formed,
+        // but succeeds on server side. However, there's no chance for
+        // client to decrypt messages encrypted by server.
+        let (cli_act_1, _msg_1) = KXHandshakeActOne::initiator(&initiator_privkey).unwrap();
+
+        // Responder doesn't fail act 1 on a bad message..
```
Comment (darosior) [inline src/noise.rs:559]: Interesting, so it doesn't validate the validity of the `e` pubkey? Glad we tested it as it's non-intuitive

Comment (darosior) [reply src/noise.rs:559]: Ok, so it's just because you use a valid key. If you use `000000000000000000000000000000..` it will fail Act 2 at the DH..

File: src/noise.rs (hunk 4 above)
Comment (darosior) [reply src/noise.rs:524]: Adressed.

File: src/noise.rs (hunk 5 above)
Comment (darosior) [reply src/noise.rs:80]: Addressed

File: src/noise.rs (hunk 1 above)
Comment (darosior) [reply src/noise.rs]: Adressed

File: src/test_utils.rs (hunk 6 above)
Comment (darosior) [reply src/test_utils.rs:28]: Adressed

File: src/noise.rs [hunk 9]
```diff
@@ -191,8 +203,8 @@ impl KKHandshakeActOne {
             .map_err(|e| Error::Noise(format!("Failed to build state for initiator: {:?}", e)))?;
 
         // Write the first message
-        let mut msg = vec![0u8; NOISE_MESSAGE_MAX_SIZE];
-        let len = state
+        let mut msg = [0u8; KK_MSG_1_SIZE];
+        let _len = state
```
[the remark this answers is not in the document]
Comment (darosior) [reply src/noise.rs:207]: Addressed

**Comment by darosior** [comment]
Addressed my comments, rebased on master, and added a small lint commit.
You can see the diff with [the `noise_old` branch](https://github.com/re-vault/revault_net/tree/noise_old) i created before processing the changes. If you can take a look to the diff to double check my review, otherwise i think it's good to go.

**Review by darosior — APPROVED** [review]
ACK 45be9d8

File: src/noise.rs (hunk 7 above)
Comment (JSwambo) [reply src/noise.rs]: Ah I see. Thanks for digging in to that!
